### ISSUE TITLE
Fully Qualified Name reference not resolved correctly

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#6553]
I'm not sure what is the exact cause of this, I guess it may be due to `Standard.Database` having a `Main.enso` file with re-exports? Just a hypothesis.

To see the issue, run the following 2 programs:

```
from Standard.Base import all
from Standard.Database import all
import Standard.Database.Connection.Postgres_Details.Postgres_Details

main =
    IO.println Postgres_Details.Postgres
```
and
```
from Standard.Base import all
from Standard.Database import all

main =
    IO.println Standard.Database.Connection.Postgres_Details.Postgres_Details.Postgres
```

They should _both_ have the same output.

However, the first one will correctly give:
```
(Postgres 'localhost' 5432 'postgres' '' Nothing Prefer Nothing)
```

but the second one, will fail with:
```
Execution finished with an error: Method `Connection` of type Main could not be found.
        at <enso> imports.main<arg-0>(imports.enso:5:16-43)
        at <enso> imports.main<arg-0>(imports.enso:5:16-60)
        at <enso> imports.main<arg-0>(imports.enso:5:16-77)
        at <enso> imports.main<arg-1>(imports.enso:5:16-86)
        at <enso> imports.main(imports.enso:5:5-86)
```

**Comment by hubertp** [comment]
This is kind of on purpose. We don't add synthetic exports (such as `Connection`) to `Main` modules because they are special.
The assumption is that library authors would do it, if needed.

**Comment by radeusgd** [comment]
> This is kind of on purpose. We don't add synthetic exports (such as `Connection`) to `Main` modules because they are special. The assumption is that library authors would do it, if needed.

I see. Well, it does however end with a very weird behaviour, that:
```
import X.Y
main = Y.Z
```
and
```
main = X.Y.Z
```
have different meanings.

Which is super confusing, isn't it? I'd expect an import and a FQN to work exactly the same way.

So maybe we need to revisit this requirement. cc: @jdunkerley @JaroslavTulach

**Comment by wdanilo** [comment]
I totally agree, this is super confusing and this breaks [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency) which is extremely bad and we should never do it. Languages that break referential transparency are providing users with very bad UX and are very confusing. So I agree with @radeusgd that we need to fix it and maintain referential transparency in all places (or as many places as we can) in the language.

@hubertp why this needs to be special? Why these two can't be the same by default? What problems would that cause?

**Comment by hubertp** [comment]
At the time when we added synthetic modules, which was a while ago, we said that there was no need to make implicit exports in `Main` modules as it was OK for them to be specified explicitly there; `Main` is the main entry point to any library component after all.

There is no problem in adding implicit exports to `Main` per se. It's just a bit cumbersome because of the special nature of how it is resolved in the compiler and the presence of existing exports and thus avoiding duplicates.

This doesn't seem like an urgent ticket though, given that we have an (easy) workaround.

**Comment by JaroslavTulach** [comment]
> There is no problem in adding implicit exports to `Main` per se. 

CCing @jdunkerley - we have chatted this week **against** having _implicit exports_. We concluded we wanted only _explicit exports_. I am not very fluent in Enso import/export design principles, so maybe our conclusion was based on wrong assumptions, but I want to mention it here.

The primary motivation for _only explicit re-exports_ was a discussion with @Akirathan working on #5936 - Pavel was afraid we would have to search modules transitively to find out who's _implicitly exporting_ a symbol. That's not very effective and (given our goal is to speed start up) we want to _resolve imports effectively_.

Whatever a resolution of this issue is, please make sure all **necessary information are available** in `Standard/Base/0.0.0-dev/.enso/cache/bindings/0.0.0-dev/Standard/Base.bindings` caches and there is _no need to parse or load `.ir` files_ just to resolve imports.

**Comment by hubertp** [comment]
One example that @jdunkerley raised, which would have undesired consequences for libs (and CB and ultimately GUI), is that among other things`Standard.Database.Internal` would become available in suggestions and that's **not** something we want to do. 
That is the reason why it would make sense to add
```
import Standard.Database.Connection
export Standard.Database.Connection
```
(and others) explicitly.

**Comment by radeusgd** [comment]
> One example that @jdunkerley raised, which would have undesired consequences for libs (and CB and ultimately GUI), is that among other things`Standard.Database.Internal` would become available in suggestions and that's **not** something we want to do. 

True, but is making imports confusing the right solution? It's just so weird that FQN behaves differently from an import.

And the CB visibility can be handled by allowing to have some metadata on the module, possibly module-level comments with a `## PRIVATE` tag to hide it. We can also just introduce a convention that the name `Internal` makes it hidden from CB (we already have the style guide suggesting to put internal stuff in modules called `Internal`, so that would be consistent).

**Comment by jdunkerley** [comment]
We can't add `Connection` unless we restructure a little, as we have a `Connection` class.

It would be good to get a feel of how much of a problem it is currently in the code base. I agree in the longer run we should make the FQN work.

**Comment by radeusgd** [comment]
Maybe indeed we just have to disallow re-exporting a type from `Main` which has a name that is clashing with its direct descendant child module?

Allowing this kind of re-exports (as is currently) will either have to make FQN somehow ambiguous OR would have to hide the module. I think both hiding and ambiguity are undesirable, so the only thing that comes to my mind is disallowing such clashing re-exports.

From the libraries perspective - what counts for us is what _types_ we want to publicly export. If any of our **internal** modules have names clashing with them (like here with `Connection`), I think we just have to rename them to avoid ambiguity - and I don't think it should be a problem to do so, right?

**Comment by radeusgd** [comment]
So IMO the resolution for this would be:

1) **Always** automatically re-export child modules (based on the directory structure).
2) If a re-export is clashing with such automatically exported module - report an error.

------

(2) will allow us to avoid ambiguity and thus we will be able to make both
```
import X.Y.Z
main = Z.W
```
and
```
main = X.Y.Z.W
```
and
```
from X.Y.Z import W
main = W
```
and
```
from X.Y import Z
main = Z.W
```
use the same resolution schemes - which I think is needed to avoid confusion - having multiple resolution schemes is really confusing.

------

And (1) is needed so that if I get hold of some element `(x : X)`, if I call `Meta.get_qualified_type_name x` and get some path, like `Standard.Database.Internal.SQLite.SQLite_Type_Mapping.SQLite_Type_Mapping`, then I should always be able to instantiate this type by using the FQN directly `Standard.Database.Internal.SQLite.SQLite_Type_Mapping.SQLite_Type_Mapping` (assuming `Standard.Database` itself is imported) or by `import Standard.Database.Internal.SQLite.SQLite_Type_Mapping.SQLite_Type_Mapping`.

------

I appreciate such _not_ re-exporting of internal modules could be a neat trick to hide implementation details, but IMO it is an emergent 'feature' that is not really desired and causes more confusion than benefit. We _do_ need some way to hide internal implementation - but that is a separate discussion and IMO it will require a different solution anyway (some kind of a `private` keyword concept).

**Comment by 4e6** [comment]
In this particular case, it is an issue with FQN resolution. I was able to reproduce it on the current example with no re-exports involved:

Given

```
$ tree
.
├── package.yaml
└── src
    ├── A.enso
    └── Main.enso
```
```
$ cat src/A.enso
type A
```
```
$ cat src/Main.enso
from Standard.Base import all
import Standard.Visualization

main =
    local.Unnamed.A.A
```

```bash
/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --no-ir-caches --run ~/enso/projects/Unnamed
Execution finished with an error: Method `A` of type Main could not be found.
        at <enso> Main.main<arg-0>(src/Main.enso:5:5-19)
        at <enso> Main.main(src/Main.enso:5:5-21)
```
[Unnamed.zip](https://github.com/enso-org/enso/files/11555807/Unnamed.zip)

**Comment by 4e6** [comment]
Also, when you remove all imports from the `Main.enso`, you get an even more interesting error
```
$ cat src/Main.enso
main =
    local.Unnamed.A.A
```
```bash
$ ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --no-ir-caches --run ~/enso/projects/Unnamed
Execution finished with an error: No_Such_Method.Error
        at <enso> Main.main<arg-0>(src/Main.enso:2:5-19)
        at <enso> Main.main(src/Main.enso:2:5-21)
```

**Comment by hubertp** [comment]
> Also, when you remove all imports from the `Main.enso`, you get an even more interesting error
> 
> ```
> $ cat src/Main.enso
> main =
>     local.Unnamed.A.A
> ```
> 
> ```shell
> $ ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --no-ir-caches --run ~/enso/projects/Unnamed
> Execution finished with an error: No_Such_Method.Error
>         at <enso> Main.main<arg-0>(src/Main.enso:2:5-19)
>         at <enso> Main.main(src/Main.enso:2:5-21)
> ```

This is expected - `to_display_text` is no longer a builtin.

**Comment by 4e6** [comment]
But from the user perspective, the program `main = local.Unnamed.A.A` does not use the `to_display_text` method

**Comment by hubertp** [comment]
> But from the user perspective, the program `main = local.Unnamed.A.A` does not use the `to_display_text` method

For reference https://github.com/enso-org/enso/issues/6451 and the reasoning described in the associated PR.

**Comment by 4e6** [comment]
Found another one:

```
from Standard.Base import all

main =
    local.Unnamed.Main.foo

foo = 42
```
Returns
```
Execution finished with an error: Method `Main` of type Main could not be found.
        at <enso> Main.main<arg-0>(src/Main.enso:4:5-22)
        at <enso> Main.main(src/Main.enso:4:5-26)
```

While the `local.Unnamed.foo` qualified name (without `Main` module) works. And `Main.foo` also works.

Not sure if it's known or not.

**Comment by 4e6** [comment]
Another one
```py
from Standard.Base import all
from Standard.Table import all

main =
    Standard.Table.Data.Aggregate_Column.Aggregate_Column.Group_By
```
```
Execution finished with an error: Method `Data` of type Main could not be found.
        at <enso> Main.main<arg-0>(src/Main.enso:5:5-23)
        at <enso> Main.main<arg-0>(src/Main.enso:5:5-40)
        at <enso> Main.main<arg-0>(src/Main.enso:5:5-57)
        at <enso> Main.main(src/Main.enso:5:5-66)
```

**Comment by JaroslavTulach** [comment]
Here in reported problems are solvable by [this diff](https://github.com/enso-org/enso/issues/9329#issuecomment-1985234571). Running:
```ruby
from Standard.Base import all
from Standard.Database import all

main =
    IO.println Standard.Database.Connection.Postgres_Details.Postgres_Details.Postgres
```
produces `(Postgres 'localhost' 5432 'postgres' '' Nothing Prefer Nothing)`.

**Comment by JaroslavTulach** [comment]
Fixed by #9539 - both programs behave the same.
